samza-core/src/main/java/org/apache/samza/coordinator/stream/CoordinatorStream.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.coordinator.stream;
+
+import org.apache.samza.config.Config;
+import org.apache.samza.metrics.MetricsRegistry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Encapsulates the creation and management of the coordinator stream producer and consumer.
+ */
+public class CoordinatorStream {
```

Comment:
What's the difference between this class and AbstractCoordinatorStreamManager? Looks like that one also maintains the producer/consumer?

Reply:
This is where it got a little confusing for me. The consumer and producer lifecycles were partially handled outside of the `AbstractCoordinatorStreamManager` and partially in it. Also, I wasn't sure what the purpose of having this abstract class other than some code re-use. The name was kind of confusing too because the subclasses are not really variations to a coordinator stream manager, but instead they are "managers that use the coordinator stream". Composition might be better? What do you think? If so, I can move all the behavior to `CoordinatorStream`. Just a thought.

Reply:
After looking at the current code, here's what I'd recommend:

1. Change AbstractCoordinatorStreamManager to a concrete class (CoordinatorStreamManager). 
3. Make CoordinatorStreamSystemConsumer and CoordinatorStreamSystemProducer private to the CoordinatorStreamManager. CoordinatorStreamManager should have two constructors: one that instantiates everything from Config (for general use), and one that accepts a SystemConsumer and SystemProducer and wraps it with the CoordinatorStreamSystemConsumer and CoordinatorStreamSystemProducer (for testing/custom lifecycle management).
3. CoordinatorStreamManager should proxy all public methods in the CSSC/CSSP.
4. LocalityManager/ChangelogManager/TaskAssignmentManager should accept a CoordinatorStreamManager instance instead of extending the abstract class.
4. Remove the CoordinatorStream class.

It looks like a fairly straightforward refactor. If you prefer to do it in a separate PR, that's OK too. @nickpan47, does this seem reasonable?